### ISSUE TITLE
Evaluate and decide on partitioning method

### ISSUE THREAD

**Issue by noahmanger** [18F/openFEC#1566]
By the end of Sprint 10, we want to have identified (and if possible, begin to implement) a method for partitioning the large tables. 

The first step is to evaluate Postgres partitioning utility to see if download times improve enough. Our goal is to get basic `sched_a` downloads to under a minute and to ensure that there are no queries that break the system.

I'm probably not capturing this the best way, so please add to this issue if anything's missing or wrong.

**Comment by ccostino** [comment]
@LindsayYoung and I had a chance to catch up about this and discussed the following:

* How partitioning in PostgreSQL works in general and what it might look like for the `ofec_sched_a` (59,092,128 records currently) and `ofec_sched_b` (27,528,889 records currently) tables.
* What the potential challenges and cons might be, such as ensuring we partition data across a suitable constraint and maintenance tasks in the future.
* What [`pg_partman`](https://github.com/keithf4/pg_partman) might do for us and its pros and cons.
* Reviewing @jmcarp's original PR: https://github.com/18F/openFEC/pull/1429
* How we want to approach this going forward.

**TL;DR**

We need to partition the `ofec_sched_a` and `ofec_sched_b` tables.  We recommend taking an iterative approach and starting with just writing the SQL first to create the partitions, trigger function, and constraint checks.  We would script what we can to help automate this for ourselves as much as possible.  Once we finish, we can evaluate the new state of these tables and what performance impacts/gains we've achieved.  At this point we can also make a decision to either release this implementation as is or immediately move on to working in `pg_partman` as what we ultimately release.

**Detailed discussion**

In short, partitioning of the 'ofec_sched_a` and `ofec_sched_b` tables should help alleviate a lot of the problems we are currently having with the size of these tables, particularly with respect to query performance, how long the update scripts take to run, and the ability to export/download data.  Given that these tables are only going to continue to grow in size, I do not think we have a choice and need to do this.

In looking at the [PostgreSQL partitioning documentation](http://www.postgresql.org/docs/9.5/static/ddl-partitioning.html) a lot more closely to understand how this works under the hood it appears to be straightforward.  [This blog post written by Hubert Lubaczewski](http://www.depesz.com/2015/06/07/partitioning-what-why-how/) goes into even greater detail in how this works (h/t to @jmcarp for sharing it with us!) and walks through a sample use-case that deals with a table containing over 666 million records and how performance is impacted during all phases of partitioning from start to finish.  What is really nice about this example is that it's written with the intent of explaining what happens when using `pg_partman` and accomplishing this while running a live application without bringing it down.

Speaking of `pg_partman`, it appears relatively straightforward enough to use as well and only does a couple of things different from the natural order of steps explained in the official documentation.  This primarily revolves around defining all indexes and constraints on the parent/master table so that the child tables that `pg_partman` creates and manages inherits the entire structure needed.  There is one caveat though in that it there is an optional dependency on the [`pg_jobmon`](https://github.com/omniti-labs/pg_jobmon) extension.  Given that we're looking at partitioning based on election cycles (every two years), this would be something we want to consider when going down the extension route to help further manage our partitions due to the date/time-based nature of the cycles.

**Evaluation**

In evaluating all of these options and approaches, including @jmcarp's implementation in Python using SQLAlchemy, one of the other things we considered was the ease of maintenance and other developers' ease of understanding, both on the 18F and FEC side.  @LindsayYoung and I are both of the mind that the simplest solution that we can go with will serve the project better in the long run and allow both us and others to maintain this part of the system more easily.  Ultimately we feel that the Python implementation, while straightforward in purpose, is not necessarily straightforward in practice.  In reviewing the code as it is (taking into account that some improvements could likely be made, though I think it's a very solid approach!), neither of us feel confident or comfortable making adjustments should we have to and we worry how that would bode for future team members as well.

We definitely like the idea of the `pg_partman` extension managing things for us, but this adds another layer into the system that we need to learn and account for with no baseline state and performance metrics to compare against.  Again, we also considered what this might mean for other folks stepping into the project in the future and trying to wrap their heads around the entire system and data set.  PostgreSQL extensions are not something we're currently familiar with and I want to make sure that I can definitely speak to using them before introducing them to other folks.

**Conclusion**

We've decided that the best way to approach this is to stay as simple as possible at first.  Creating a small collection of SQL scripts to take care of the initial partitioning and then handling the trigger(s) and constraint checks will be straightforward to create and implement.  It will also provide us with the baseline state and understanding of performance impact that we're looking for before going straight into relying on `pg_partman`.

Once we have this working we can then decide if we want to release the implementation as is and then continue working toward utilizing the extension, or simply test internally for a bit and press on with a `pg_partman` implementation that we ultimately release.  We feel that this is the best way to introduce what we're doing to others working in the project and ensuring that what we're working toward makes sense and has the outcome that we expect.

**References and additional information**

http://www.postgresql.org/docs/9.5/static/ddl-partitioning.html
http://www.depesz.com/2015/04/02/waiting-for-9-5-allow-foreign-tables-to-participate-in-inheritance/
http://pgxn.org/dist/pg_partman/doc/pg_partman.html
http://pgxn.org/dist/pg_jobmon/doc/pg_jobmon.html
https://github.com/keithf4/pg_partman
https://github.com/omniti-labs/pg_jobmon
http://www.postgresql.org/docs/current/static/sql-createextension.html
https://github.com/18F/openFEC/pull/1429

Any thoughts, questions, criticisms, etc. are welcome to this.  We definitely want to make sure that we're on the right track!

**Comment by noahmanger** [comment]
I have nothing to add to the technical discussion, but just want to +1 the way you broke this out into smaller tasks. And thanks for the thorough write-up of the considerations and research. Have you circulated this with other devs who might be able to weigh in?

**Comment by ccostino** [comment]
Thanks, @noahmanger!  I have not yet, but thank you for reminding me because I wanted to catch up with @catherinedevlin about this (if you have a moment to do so!). :-)

**Comment by jmcarp** [comment]
I can see writing scripts for each child table by hand as a quick and dirty benchmark, but I wouldn't recommend it for anything else. If you hand-write all the scripts, you're talking about at least ten tables (two large tables in two-year chunks with data going back to IIRC 2007 or earlier). That's going to be effortful and error-prone to maintain--every time you change one table, you have to change them all.

Here's what I'd recommend.

* Figure out what your goals are. Faster data exports? Faster inserts? Speeding up certain queries?
* Profile those things using the current schemas. Write down how long your reference tasks take.
* Partition *one* of the tables by hand. Don't send a pull request--this is just for benchmarking.
* Profile your reference tasks again.
* If everything got faster, great! Throw away your profiling code and either use `pg_partman` or some python scripts to manage the partitions.

**Comment by ccostino** [comment]
@jmcarp, as I've been thinking about this more this has become my concern as well, especially after speaking with @catherinedevlin this morning about some things related to partitioning.  One of the things we discussed was getting a read replica in place and partitioning in there to see what happens without affecting everything else.

I think what you've outlined makes a lot of sense, and would allow us to experiment and test our theories with the safety net of going with a more robust solution once we're ready with something based on our findings.

**Comment by noahmanger** [comment]
@ccostino are those all things that will happen this sprint? Or should we break them out into separate issues?

**Comment by ccostino** [comment]
@noahmanger: Definitely separate issues - I need to get @catherinedevlin some SQL to help investigate.  @jmcarp's recommendations are very sensible approaches to the steps and findings @LindsayYoung and I agreed upon above after discussing this and performing some research.  Basically what this means is we won't release any interim testing/experimentation we do, we'll just keep it for ourselves for benchmarking/profiling purposes.

We know our primary goal is to enable reasonably fast downloads to these tables, so that is going to be our focus with this.  I think the first step reviewing the current state of one of the tables to see if there's anything else we need to consider or address, then setup a read replica for ourselves to experiment with any modifications and the actual partitioning (manually).

I'm happy to write the issues up for us and work with you to get them scheduled correctly!

**Comment by ccostino** [comment]
So a quick update on this after digging deeper into what's possible:  we're going to have to manage partitioning ourselves on Amazon RDS.  Neither `pg_partman` or `pg_jobmon` (optional, but recommended by the `pg_partman` author) is in the [list of supported extensions](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html#PostgreSQL.Concepts.General.FeatureSupport).  I double checked by running the `SHOW rds.extensions;` command to be sure, this was the output:

```
tree_gin,btree_gist,chkpass,citext,cube,dblink,dict_int,dict_xsyn,earthdistance,fuzzystrmatch,hstore,intagg,intarray,ip4r,isn,ltree,pgcrypto,pgrowlocks,pgstattuple,pg_buffercache,pg_prewarm,pg_stat_statements,pg_trgm,plcoffee,plls,plperl,plpgsql,pltcl,plv8,postgis,postgis_tiger_geocoder,postgis_topology,postgres_fdw,sslinfo,tablefunc,test_parser,tsearch2,unaccent,uuid-ossp
```

I did some additional digging to see if it's possible to install your own extensions, [sadly it's not](http://dba.stackexchange.com/a/111481) in this case, and there is already an [open feature request](https://forums.aws.amazon.com/thread.jspa?messageID=542044&#542044) for it going back nearly two years.

**Comment by noahmanger** [comment]
Bummer. So where does that leave us?

**Comment by ccostino** [comment]
@noahmanger: As far as the partitioning is concerned, it brings us back to the approach @jmcarp was taking: executed manually but managed via scripts of some sort.  I've also been looking into some Postgres performance tuning and there are several other things I'd like to investigate, too.  I know I haven't written the follow-up issues to break this apart yet, sorry.  I've been clearing my plate of other things while researching!

**Comment by ccostino** [comment]
At this juncture, we now know that we have to manage the table partitions ourselves, and after more discussion we feel @jmcarp's original approach is the best way of doing so.  In settling on this approach we're going to close this issue and create new ones that represent the next tasks of benchmarking and having conversations with the folks at FEC to review what we plan to do and ensure our goals align.

Thanks, everyone!
